test/e2e/memtierd.test-suite/no-policy/n4c16/test01-mover/code.var.sh

```diff
@@ -49,15 +62,81 @@ mover_min_max_seconds=(
     "1 2"
 )
 
-for target_numa_node in {0..3}
-do
-  MEMTIERD_YAML=""
-  memtierd-start
-  memtierd-command "pages -pid ${MEME_PID}"
-  memtierd-command "mover ${mover_conf[$target_numa_node]} -pages-to ${target_numa_node}"
-  echo "waiting 700M+ to be moved to ${target_numa_node}"
-  match-moved "${target_numa_node}\:0\.7[0-9][0-9]" ${target_numa_node} ${mover_min_max_seconds[$target_numa_node]}
-  memtierd-stop
+echo -e "\n=== scenario 1: test moving memories among numa nodes ===\n"
+
+# Test that 700M+ is moved to numa node {0..3} from a process
+MEME_BS=720M MEME_BWC=1 MEME_BWS=720M memtierd-meme-start
+
+for target_numa_node in {0..3}; do
+    MEMTIERD_YAML=""
+    memtierd-start
+    memtierd-command "pages -pid ${MEME_PID}"
+    memtierd-command "mover ${mover_conf[$target_numa_node]} -pages-to ${target_numa_node}"
+    echo "waiting 700M+ to be moved to ${target_numa_node}"
+    match-moved "${target_numa_node}\:0\.7[0-9][0-9]" ${target_numa_node} ${mover_min_max_seconds[$target_numa_node]}
+    memtierd-stop
+done
+
+memtierd-meme-stop
+
+echo -e "\n=== scenario 2: test swapping out memories ===\n"
+
+echo -e "\n=== scenario 2.1: test swapping out memories for one process with various mover configurations ===\n"
+
+for i in {0..3}; do
+    # Test that 700M+ is swapped out
+    MEME_BS=1G MEME_BWC=1 MEME_BWS=280M memtierd-meme-start
+    MEMTIERD_YAML=""
+    memtierd-start
+
+    # the 1st time swapping out with -mover option
+    memtierd-command "mover ${mover_conf[$i]}\nswap -out -pids ${MEME_PID} -mover\nmover -wait"
+    match-pid-pageout ${MEME_PID} "7[0-9][0-9]" ${mover_min_max_seconds[$i]}
+
+    # swapping in all the memory which have been swapped out above
+    # thus, 0-9 MB will be swapped out after swapping in option
+    memtierd-command "swap -in -pids ${MEME_PID}"
+    match-pid-pageout ${MEME_PID} "[0-9]" 0 1
+
+    # the 2nd time swapping out without -mover option
+    memtierd-command "swap -out -pids ${MEME_PID}"
+    match-pid-pageout ${MEME_PID} "7[0-9][0-9]" 0 1
+
+    memtierd-meme-stop
+    memtierd-stop
+done
+
+echo -e "\n=== scenario 2.2: test swapping out memories for multiple processes ===\n"
+
+# Start 3 meme, each should have 700M+ swapped out
+for j in {0..2}; do
+    MEME_BS=1G MEME_BWC=1 MEME_BWS=280M memtierd-meme-start
+    PID_VAR="MEME${j}_PID"
+    eval "${PID_VAR}=${MEME_PID}"
+done
+MEME_PIDS=("${MEME0_PID}" "${MEME1_PID}" "${MEME2_PID}")
+
+MEMTIERD_YAML=""
+memtierd-start
+
+# the 1st time swapping out with -mover option for the three meme processes
+memtierd-command "mover -config {\"IntervalMs\":20,\"Bandwidth\":500}\nswap -out -pids ${MEME0_PID},${MEME1_PID},${MEME2_PID} -mover\nmover -wait"
+# As mover handles the multiple pids' tasks one by one, thus, triple the max_seconds and set the min_seconds as 0
+for MEME_PID in "${MEME_PIDS[@]}"; do
+    match-pid-pageout "${MEME_PID}" "7[0-9][0-9]" 0 6
```

Comment:
Configured mover bandwidth, 500 MB/s, is more than we can realistically expect to get in a virtual machine. Therefore the bandwidth is probably not throttling anything, and we might miss an error where mover is not throttling properly.

Could we change this so that we would use relatively small memory sizes in memes, and use significantly smaller mover bandwidth. For instance 10 MB/s or even less. Then we could use match-pid-pageout min time somewhere in the range of 3-6 seconds.

My goal is that if there is a bug that omits the mover, we would always get a failure here, because swapping would be too fast. On the other hand, if the mover calculates MB/s somehow wrong, we would detect that, too, because swapping might be significantly slower than expected.

Reply:
Fixed~